src/transaction/types/tx_pointer.rs

```diff
@@ -0,0 +1,155 @@
+use fuel_types::bytes::{SizedBytes, WORD_SIZE};
+
+use core::{fmt, str};
+
+#[cfg(feature = "std")]
+use fuel_types::bytes;
+
+#[cfg(feature = "std")]
+use std::io;
+
+#[cfg(feature = "random")]
+use rand::{
+    distributions::{Distribution, Standard},
+    Rng,
+};
+
+/// Identification of unspend transaction output.
+#[derive(Debug, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct TxPointer {
+    /// Block height
+    block_height: u32,
+    /// Transaction index
+    tx_index: u16,
+}
+
+impl TxPointer {
+    pub const LEN: usize = 2 * WORD_SIZE;
+
+    pub const fn new(block_height: u32, tx_index: u16) -> Self {
+        Self {
+            block_height,
+            tx_index,
+        }
+    }
+
+    pub const fn block_height(&self) -> u32 {
+        self.block_height
+    }
+
+    pub const fn tx_index(&self) -> u16 {
+        self.tx_index
+    }
+}
+
+#[cfg(feature = "random")]
+impl Distribution<TxPointer> for Standard {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> TxPointer {
+        TxPointer::new(rng.gen(), rng.gen())
+    }
+}
+
+impl fmt::LowerHex for TxPointer {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{:08x}{:04x}", self.block_height, self.tx_index)
+    }
+}
+
+impl fmt::UpperHex for TxPointer {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{:08X}{:04X}", self.block_height, self.tx_index)
+    }
+}
+
+impl str::FromStr for TxPointer {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        const ERR: &str = "Invalid encoded byte";
+
+        if s.len() != 12 {
+            return Err(ERR);
+        }
+
+        let block_height = u32::from_str_radix(&s[..8], 16).map_err(|_| ERR)?;
+        let tx_index = u16::from_str_radix(&s[8..12], 16).map_err(|_| ERR)?;
+
+        Ok(Self::new(block_height, tx_index))
+    }
+}
+
+impl SizedBytes for TxPointer {
+    fn serialized_size(&self) -> usize {
+        Self::LEN
+    }
+}
+
+#[cfg(feature = "std")]
+impl io::Write for TxPointer {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if buf.len() < Self::LEN {
+            return Err(bytes::eof());
+        }
+
+        // Safety: buf len is checked
+        let (block_height, buf) = unsafe { bytes::restore_word_unchecked(buf) };
+        let (tx_index, _) = unsafe { bytes::restore_word_unchecked(buf) };
+
+        self.block_height = block_height as u32;
+        self.tx_index = tx_index as u16;
```

Comment:
Is there a possibility that these `as` casts are sinlently truncating?

Reply:
The original thought is that it might be overkill to check that since block height is `u32` everywhere in the stack, as `tx_index` is `u16`. Our canonical serialization only encode/decode words/`u64`, so we need to truncate

But you're right; better safe than sorry: https://github.com/FuelLabs/fuel-tx/pull/170/commits/1368be316e15c67e01a9fb43fff7f4324b103fc9